### ISSUE TITLE
Feature Request: Integrating a developer environment into a Mesh

### ISSUE THREAD

**Issue by lavignes** [aws/aws-app-mesh-roadmap#184]
Developers running/testing their apps locally (i.e. using tools like docker-compose) in their own environments may want to be able to integrate with a Mesh running in a VPC.

I'm proposing we make an example of using Client VPN to add a multi-container local docker environment as a virtual node to an existing mesh.

**Comment by nathanpeck** [comment]
I like this feature request. I'd say the gold standard for experience here would be something like telepresence: https://www.telepresence.io/

Provide a toolchain that makes it easy to launch a proxy in the mesh inside my AWS VPC, exposed to the public internet so it can be connected to a local proxy on my desktop. Then I can run my container on my local machine, it talks to a local Envoy which talks to the remote inside my VPC, which can send traffic to a real production virtual node

**Comment by janaka** [comment]
In our current set up, when an engineer runs a microservice on their laptop in debug mode, it gets registered as a node in service discovery in our test dev environment in AWS. So from service discovery and interservice comps point of view the instance of the app on the devs machine is treated the same as all other instances of microservices running in the dev test environment in the AWS VPC. The engineer would be on the corp VPN (or physically on the office network) for the network path to work. Today we use Eureka for service discovery and RPC over HTTP for interservice comms. 

Ideally if/when we move to App Mesh and CloudMap we'd want to be able to provide a similar experience. To do that we probably need to be able to register a locally running app (native or in a container) with Cloud Map and App Mesh, be able to have App Mesh manage the local Envoy instance, and of course network traffic routing. I'm probably missing some details here but I think those are the basics. 

Our engineering team are on a mix of macOS and Windows with a sprinkle of Linux.

**Comment by apottere** [comment]
This is probably the biggest blocker for us adopting AppMesh for our microservices - to the point that we might just move all of our ECS services into EKS to take advantage of telepresence.  Almost all of our services are dependent on other services and data sources that we don't want to or can't run locally.  Right now we use a VPN and do "service discovery" by putting ALB hostnames in Parameter Store, which means our local applications can resolve and call applications running in our dev environment.

If I understand correctly, it should be possible to write a script that sets up a local Envoy instance, connects it (through some sort of VPN) to AppMesh or an Envoy instance in AWS connected to the mesh, and use that to somehow proxy connections for our local applications.  Figuring out _how_ to set that up is another thing entirely.  It seems like the aws cli should have a telepresence-esque command that can do that all for you.

**Comment by ssypi** [comment]
@apottere Did you find a good solution for this? I've been testing running a socks proxy through SSH connection to an EC2 instance that is running envoy connected to the mesh. It feels a bit dirty solution, but it kinda works. We can use proxychains to run our code (eg. "proxychains npm start") and can do dns resolution and connecting to mesh services through the proxy. If needed you can also run ssh remote port forwarding to enable other mesh services to connect to the locally running dev server. Just wondering what you ended up with and whether you came up with anything better.

**Comment by apottere** [comment]
@ssypi No, we never adopted AppMesh because we didn't feel it was worth the effort to figure this out if EKS was on the roadmap anyway.

**Comment by EladG77** [comment]
Any news about this?

**Comment by paulohenriqu** [comment]
Any news about that?
A telepresence like solution would be great. That's blocking us also to adopt AppMesh.
